Password Validator Improvements

JVickery-TBS opened the pull request:
Adds `IValidators` implementation. Adds a bunch of configs to control monkey-patching, password length, and password complexity.

JVickery-TBS committed:
feat(logic): validators;

- Moved monkey patching excessive schemas to `IValidator` implementation.
- Moved monkey patching schema old user name to chained action.
- Added more config options for password requirements.

JVickery-TBS committed:
feat(logic): config;

- Added more config options for controlling monkey patching vs IValidators.

JVickery-TBS committed:
feat(misc): changelogs;

- Added change log files.

wardi commented on ckanext/security/logic/action.py:
```diff
@@ -67,6 +69,10 @@ def user_update(up_func, context, data_dict):
+    # (canada fork only): update the user update form schema for username field
+    # TODO: upstream contrib??
+    if asbool(config.get('ckanext.security.use_ivalidators', False)):
+        context['schema']['name'].append(validators.old_username_validator)
```
use `get_validator()` here so that other plugins can override the definition?

JVickery-TBS replied:
put `old_username_validator` in plugin's `get_validators` and used toolkit here now.

wardi commented on ckanext/security/plugin/__init__.py:
```diff
@@ -27,30 +28,47 @@ class CkanSecurityPlugin(MixinPlugin, p.SingletonPlugin):
+            core_schema.user_edit_form_schema = \
+                ext_schema.user_edit_form_schema
+            core_schema.default_update_user_schema = \
+                ext_schema.default_update_user_schema
```
Is there a reason to keep the old monkey patching and not use IValidators?

JVickery-TBS replied:
@wardi yeah, I don't think so. I don't know why I had the idea to put it behind a config option. It is best to remove it especially if Core adds anything to the user schemas and security does not have it.

For instance, I added that new "cannot modify your own sysadmin privs" to the user schema, and the security extension does not have that in the monkey patch.

wardi commented on ckanext/security/validators.py:
```diff
@@ -5,13 +5,16 @@
+    'Your password must consist of at least three of the following character sets: '
+    'uppercase characters, lowercase characters, digits, punctuation & special characters.'
 )
+NZISM_COMPLIANT = asbool(config.get('ckanext.security.nzism_compliant_passwords', True))
```
loading config at module level means we can't write tests that exercise different values for this option in the normal way

JVickery-TBS replied:
Realllly. okay that is good to know. I will just move the definitions inside the validator method then

wardi replied:
yeah, nothing's lazy in python. If it's at the module level then it's executed only once at import time.

JVickery-TBS committed:
fix(schema): config in validator;

- Moved config gets inside validator method.
- Registered `old_username_validator` as an actual validator.
- Removed monkey patching.
